<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Bootstrap motifbreakR results for statistical analysis on TFBS-disrupting SNPs in upstream regions of a set of genes of interest.">
<title>Bootstrapping and statistical analysis for TFBS-disrupting SNPs • tfboot</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Bootstrapping and statistical analysis for TFBS-disrupting SNPs">
<meta property="og:description" content="Bootstrap motifbreakR results for statistical analysis on TFBS-disrupting SNPs in upstream regions of a set of genes of interest.">
<meta property="og:image" content="https://colossal-compsci.github.io/tfboot/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">tfboot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/intro.html">Statistical analysis of TFBS disruption with tfboot</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/colossal-compsci/tfboot/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="tfboot-">tfboot 
<a class="anchor" aria-label="anchor" href="#tfboot-"></a>
</h1>
</div>
<!-- badges: start -->

<p>The goal of tfboot is to facilitate statistical analysis of SNPs disrupting transcription factor binding sites (TFBS) using bootstrapping resampling to create empirical null distributions.</p>
<!-- FIXME: Update with citation:  -->
<!-- If you use tfboot, please consider citing the [paper](https://www.biorxiv.org/): Turner, S.D., et al. (2023). tfboot: Bootstrapping and statistical analysis for transcription factor binding site-disrupting variants in gene sets. bioRxiv 2023.MM.DD.XXXXXX. doi:FIXME. -->
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install tfboot from <a href="https://github.com/colossal-compsci/tfboot" class="external-link">GitHub</a> with the code below. To install only tfboot:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"colossal-compsci/tfboot"</span><span class="op">)</span></span></code></pre></div>
<p>To install required and suggested packages, including motifbreakR and those needed to build the vignette:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"colossal-compsci/tfboot"</span>,</span>
<span>                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                         dependencies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Imports"</span>, <span class="st">"Suggests"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>See the <a href="https://colossal-compsci.github.io/tfboot">pkgdown documentation</a> for an <a href="https://colossal-compsci.github.io/tfboot/articles/intro.html">introductory vignette</a> and function documentation.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>Let’s use an example from the vignette. First, let’s load some pre-baked motifbreakR results. <code>mbres</code> is a set of motifbreakR results run on SNPs in the 5kb promoter region of a random selection of 5 genes. <code>mball</code> is the precomputed set of motifbreakR results run on SNPs in the promoter region of <em>all</em> genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/colossal-compsci/tfboot" class="external-link">tfboot</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbres</span> <span class="op">&lt;-</span> <span class="va">vignettedata</span><span class="op">$</span><span class="va">mbres</span></span>
<span><span class="va">mbres</span></span>
<span><span class="co">#&gt; # A tibble: 1,335 × 10</span></span>
<span><span class="co">#&gt;    gene_id   SNP_id            tf    pctRef pctAlt scoreRef scoreAlt effect alleleDiff alleleEffectSize</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;chr&gt;             &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 100316005 chr33:7472152_G/C ACE2   0.986  0.783     4.85     3.87 strong     -0.980          -0.199 </span></span>
<span><span class="co">#&gt;  2 100316005 chr33:7472176_T/G ADR1   0.759  0.917     3.34     4.01 weak        0.673           0.154 </span></span>
<span><span class="co">#&gt;  3 100316005 chr33:7471139_A/C AFT2   0.964  0.878     5.50     5.02 weak       -0.480          -0.0843</span></span>
<span><span class="co">#&gt;  4 100316005 chr33:7472176_T/G AFT2   0.902  0.740     5.15     4.24 strong     -0.911          -0.160 </span></span>
<span><span class="co">#&gt;  5 100316005 chr33:7473834_G/T AFT2   0.879  0.705     5.02     4.04 strong     -0.980          -0.172 </span></span>
<span><span class="co">#&gt;  6 100316005 chr33:7472152_G/C AGL42  0.859  0.693     5.14     4.15 strong     -0.986          -0.165 </span></span>
<span><span class="co">#&gt;  7 100316005 chr33:7471139_A/C AGL55  0.695  0.861     3.97     4.91 strong      0.939           0.165 </span></span>
<span><span class="co">#&gt;  8 100316005 chr33:7472830_A/T ALX3   0.972  0.776     4.23     3.43 strong     -0.806          -0.185 </span></span>
<span><span class="co">#&gt;  9 100316005 chr33:7473857_T/C ALX3   0.919  0.773     4.02     3.42 weak       -0.600          -0.138 </span></span>
<span><span class="co">#&gt; 10 100316005 chr33:7473583_G/A ARG80  0.731  0.961     3.16     4.15 strong      0.993           0.230 </span></span>
<span><span class="co">#&gt; # ℹ 1,325 more rows</span></span>
<span><span class="va">mball</span> <span class="op">&lt;-</span> <span class="va">vignettedata</span><span class="op">$</span><span class="va">mball</span></span>
<span><span class="va">mball</span></span>
<span><span class="co">#&gt; # A tibble: 12,689 × 10</span></span>
<span><span class="co">#&gt;    gene_id   SNP_id            tf        pctRef pctAlt scoreRef scoreAlt effect alleleDiff alleleEffectSize</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;chr&gt;             &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 100315917 chr33:5706617_C/T ARG80      0.979  0.749     4.23     3.24 strong     -0.993           -0.230</span></span>
<span><span class="co">#&gt;  2 100315917 chr33:5706617_C/T ARG81      0.991  0.825     5.89     4.90 strong     -0.982           -0.165</span></span>
<span><span class="co">#&gt;  3 100315917 chr33:5703167_G/T ARGFX      0.781  0.889     5.06     5.74 weak        0.677            0.105</span></span>
<span><span class="co">#&gt;  4 100315917 chr33:5706617_C/T ARR1       0.634  0.880     2.53     3.42 strong      0.892            0.231</span></span>
<span><span class="co">#&gt;  5 100315917 chr33:5707492_C/T AT1G19040  0.882  0.779     7.60     6.73 strong     -0.861           -0.100</span></span>
<span><span class="co">#&gt;  6 100315917 chr33:5704406_C/G AT3G46070  0.784  0.892     5.13     5.79 weak        0.666            0.103</span></span>
<span><span class="co">#&gt;  7 100315917 chr33:5706617_C/T ATHB-12    0.781  0.890     4.12     4.67 weak        0.552            0.106</span></span>
<span><span class="co">#&gt;  8 100315917 chr33:5706617_C/T ATHB-13    0.721  0.868     4.98     5.98 strong      0.993            0.145</span></span>
<span><span class="co">#&gt;  9 100315917 chr33:5706617_C/T ATHB-16    0.813  0.966     5.32     6.32 strong      0.998            0.153</span></span>
<span><span class="co">#&gt; 10 100315917 chr33:5706617_C/T ATHB-4     0.778  0.931     4.98     5.94 strong      0.963            0.151</span></span>
<span><span class="co">#&gt; # ℹ 12,679 more rows</span></span></code></pre></div>
<p>Summarize motifbreakR results on our 5 genes of interest. This shows us the actual values for the number of SNPs in the upstream regions of these genes, and summary statistics on the allele differences, effect sizes, etc. See the vignette and <code><a href="reference/mb_summarize.html">?mb_summarize</a></code> for details.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbsmry</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mb_summarize.html">mb_summarize</a></span><span class="op">(</span><span class="va">mbres</span><span class="op">)</span></span>
<span><span class="va">mbsmry</span></span>
<span><span class="co">#&gt; # A tibble: 1 × 7</span></span>
<span><span class="co">#&gt;   ngenes nsnps nstrong alleleDiffAbsMean alleleDiffAbsSum alleleEffectSizeAbsMean alleleEffectSizeAbsSum</span></span>
<span><span class="co">#&gt;    &lt;int&gt; &lt;int&gt;   &lt;int&gt;             &lt;dbl&gt;            &lt;dbl&gt;                   &lt;dbl&gt;                  &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1      5  1335     950             0.798            1066.                   0.155                   206.</span></span></code></pre></div>
<p>Bootstrap resample motifbreakR results for all genes. Resample sets of 5 genes 250 times.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">mbboot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mb_bootstrap.html">mb_bootstrap</a></span><span class="op">(</span><span class="va">mball</span>, ngenes<span class="op">=</span><span class="fl">5</span>, boots <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span>
<span><span class="va">mbboot</span><span class="op">$</span><span class="va">bootwide</span></span>
<span><span class="co">#&gt; # A tibble: 250 × 9</span></span>
<span><span class="co">#&gt;     boot genes                                    ngenes nsnps nstrong alleleDiffAbsMean alleleDiffAbsSum alleleEffectSizeAbsMean alleleEffectSizeAbsSum</span></span>
<span><span class="co">#&gt;    &lt;int&gt; &lt;chr&gt;                                     &lt;int&gt; &lt;int&gt;   &lt;int&gt;             &lt;dbl&gt;            &lt;dbl&gt;                   &lt;dbl&gt;                  &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1     1 431304;426469;100315917;396544;374124         5  1062     750             0.797             847.                   0.155                   164.</span></span>
<span><span class="co">#&gt;  2     2 426183;395959;431304;429501;396170            5  1244     914             0.808            1005.                   0.157                   195.</span></span>
<span><span class="co">#&gt;  3     3 102465361;426183;396544;426469;429172         5   886     655             0.812             720.                   0.149                   132.</span></span>
<span><span class="co">#&gt;  4     4 396007;407779;693245;429501;100498692         5  1514    1139             0.815            1234.                   0.153                   232.</span></span>
<span><span class="co">#&gt;  5     5 426883;396544;408041;426183;426469            5   925     623             0.787             728.                   0.147                   136.</span></span>
<span><span class="co">#&gt;  6     6 425059;429035;100529062;396007;425613         5  1350    1026             0.821            1109.                   0.155                   209.</span></span>
<span><span class="co">#&gt;  7     7 408042;426880;100498692;425499;426885         5   823     627             0.823             677.                   0.157                   129.</span></span>
<span><span class="co">#&gt;  8     8 396170;425058;426886;395587;396045            5  1262     910             0.805            1016.                   0.147                   185.</span></span>
<span><span class="co">#&gt;  9     9 102466833;426183;100529061;396045;395959      5  1263     963             0.822            1038.                   0.153                   193.</span></span>
<span><span class="co">#&gt; 10    10 429035;408042;100529062;100529061;425613      5  1289     987             0.823            1061.                   0.154                   198.</span></span>
<span><span class="co">#&gt; # ℹ 240 more rows</span></span>
<span><span class="va">mbboot</span><span class="op">$</span><span class="va">bootdist</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span><span class="co">#&gt;   metric                  bootdist   </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                   &lt;list&gt;     </span></span>
<span><span class="co">#&gt; 1 alleleDiffAbsMean       &lt;dbl [250]&gt;</span></span>
<span><span class="co">#&gt; 2 alleleDiffAbsSum        &lt;dbl [250]&gt;</span></span>
<span><span class="co">#&gt; 3 alleleEffectSizeAbsMean &lt;dbl [250]&gt;</span></span>
<span><span class="co">#&gt; 4 alleleEffectSizeAbsSum  &lt;dbl [250]&gt;</span></span>
<span><span class="co">#&gt; 5 nsnps                   &lt;dbl [250]&gt;</span></span>
<span><span class="co">#&gt; 6 nstrong                 &lt;dbl [250]&gt;</span></span></code></pre></div>
<p>Compare the values from our five genes of interest to the empirical null distribution from bootstrap resampling.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bootstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mb_bootstats.html">mb_bootstats</a></span><span class="op">(</span><span class="va">mbsmry</span>, <span class="va">mbboot</span><span class="op">)</span></span>
<span><span class="va">bootstats</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 5</span></span>
<span><span class="co">#&gt;   metric                      stat bootdist     bootmax     p</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                      &lt;dbl&gt; &lt;list&gt;         &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 nsnps                   1335     &lt;dbl [250]&gt; 1597     0.136</span></span>
<span><span class="co">#&gt; 2 nstrong                  950     &lt;dbl [250]&gt; 1210     0.188</span></span>
<span><span class="co">#&gt; 3 alleleDiffAbsMean          0.798 &lt;dbl [250]&gt;    0.831 0.836</span></span>
<span><span class="co">#&gt; 4 alleleDiffAbsSum        1066.    &lt;dbl [250]&gt; 1304.    0.148</span></span>
<span><span class="co">#&gt; 5 alleleEffectSizeAbsMean    0.155 &lt;dbl [250]&gt;    0.163 0.456</span></span>
<span><span class="co">#&gt; 6 alleleEffectSizeAbsSum   206.    &lt;dbl [250]&gt;  251.    0.128</span></span></code></pre></div>
<p>Visualize the results:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_bootstats.html">plot_bootstats</a></span><span class="op">(</span><span class="va">bootstats</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-plot_bootstats-1.png" width="100%"></p>
<p>See <code><a href="articles/intro.html">vignette("intro", package="tfboot")</a></code> or the articles on the <a href="https://colossal-compsci.github.io/tfboot">pkgdown documentation website</a> for more.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/colossal-compsci/tfboot/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/colossal-compsci/tfboot/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tfboot</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stephen Turner <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9140-9028" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/colossal-compsci/tfboot/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/colossal-compsci/tfboot/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://colossal-compsci.github.io/tfboot"><img src="https://img.shields.io/badge/documentation-pkgdown-9164cc.svg" alt="pkgdown"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Stephen Turner.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
