---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
options(tibble.width = Inf)
options(width = 500)
```

```{r build, include=FALSE, eval=FALSE}
devtools::build_readme()
devtools::build_vignettes()
devtools::install(build_vignettes = TRUE, upgrade=FALSE)
devtools::build_manual(pkg=".", path=".")
file.rename(sprintf("tfboot_%s.pdf", packageVersion("tfboot")), "tfboot.pdf")
postdoc::render_package_manual("tfboot", outdir=".")
```


# tfboot

<!-- badges: start -->
<!-- badges: end -->

The goal of tfboot is to facilitate statistical analysis of SNPs disrupting transcription factor binding sites (TFBS) using bootstrapping resampling to create empirical null distributions.

## Installation

You can install the development version of tfboot from [GitHub](https://github.com/) with the code below. You'll need to replace `gh_pat` with your [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token). 

```{r, eval=FALSE}
# First install devtools if you don't have it already
# install.packages("devtools")

# Set your personal access token
gh_pat="12345-abcde-CHANGEME-67890-xyz"

# Install directly from GitHub:
devtools::install_github(repo = "https://github.com/colossal-compsci/tfboot", 
                         auth_token = gh_pat, 
                         upgrade = FALSE, 
                         build_vignettes = TRUE, 
                         force = TRUE, 
                         dependencies = c("Imports", "Suggests"))
```

## Documentation

See the full introductory vignette with:

```{r, eval=FALSE}
vignette("intro", package="tfboot")
```

## Example

Let's use an example from the vignette. First, let's load some pre-baked motifbreakR results. `mbres` is a set of motifbreakR results run on SNPs in the 5kb promoter region of a random selection of 5 genes. `mball` is the precomputed set of motifbreakR results run on SNPs in the promoter region of _all_ genes. 

```{r, include=FALSE}
library(tibble)
library(tfboot)
```

```{r}
library(tfboot)
```

```{r}
mbres <- vignettedata$mbres
mbres
mball <- vignettedata$mball
mball
```

Summarize motifbreakR results on our 5 genes of interest. This shows us the actual values for the number of SNPs in the upstream regions of these genes, and summary statistics on the allele differences, effect sizes, etc. See the vignette and `?mb_summarize` for details.

```{r}
mbsmry <- mb_summarize(mbres)
mbsmry
```

Bootstrap resample motifbreakR results for all genes. Resample sets of 5 genes 250 times.

```{r}
set.seed(42)
mbboot <- mb_bootstrap(mball, ngenes=5, boots = 250)
mbboot$bootwide
mbboot$bootdist
```

Compare the values from our five genes of interest to the empirical null distribution from bootstrap resampling.

```{r}
bootstats <- mb_bootstats(mbsmry, mbboot)
bootstats
```

Visualize the results:

```{r plot_bootstats, fig.width=7, fig.height=5}
plot_bootstats(bootstats)
```

See `vignette("intro", package="tfboot")` for more.
