<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tfboot">
<title>Statistical analysis of TFBS disruption with tfboot • tfboot</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Statistical analysis of TFBS disruption with tfboot">
<meta property="og:description" content="tfboot">
<meta property="og:image" content="https://colossal-compsci.github.io/tfboot/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tfboot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/intro.html">Statistical analysis of TFBS disruption with tfboot</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/colossal-compsci/tfboot/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Statistical analysis of TFBS disruption with tfboot</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/colossal-compsci/tfboot/blob/HEAD/vignettes/intro.Rmd" class="external-link"><code>vignettes/intro.Rmd</code></a></small>
      <div class="d-none name"><code>intro.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h3>
<p>The <a href="https://bioconductor.org/packages/motifbreakR/" class="external-link">motifbreakR</a>
package provides functions to assess how single nucleotide polymorphisms
(SNPs) disrupt transcript factor binding sites (TFBS). The tfboot
package provides helper functions to aid in assessing the statistical
significance of TFBS disruption in <em>interval sets</em>, typically
defined by -5kb upstream promoter regions on genes of interest.</p>
<p>An example question might be: we have these 123 genes of interest
that we think are important for XYZ phenotype. Can you assess the degree
to which SNPs in the promoter region of these genes disrupt TFBS? Is
this disruption in these 123 genes statistically significant as compared
to SNPs from randomly selected sets of the same number of genes?</p>
</div>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>To run this vignette, you’ll need the chicken BSGenome+TxDb+org.db
packages, plyranges, and motifbreakR packages, which can be installed
from bioconductor. You’ll also need to install this tfboot package,
which you can do from source, or directly from GitHub.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BSgenome.Ggallus.UCSC.galGal6"</span>, </span>
<span>                       <span class="st">"org.Gg.eg.db"</span>, </span>
<span>                       <span class="st">"TxDb.Ggallus.UCSC.galGal6.refGene"</span>, </span>
<span>                       <span class="st">"plyranges"</span>, </span>
<span>                       <span class="st">"motifbreakR"</span><span class="op">)</span>, update<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"https://github.com/colossal-compsci/tfboot"</span>, upgrade<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>First, load the libraries you need for this analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Ggallus.UCSC.galGal6</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Ggallus.UCSC.galGal6.refGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Gg.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">motifbreakR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/colossal-compsci/tfboot" class="external-link">tfboot</a></span><span class="op">)</span></span></code></pre></div>
<p>Use <code><a href="../reference/read_vcf.html">read_vcf()</a></code> to read in your VCF. The example used
here is a VCF with variants from chicken (<em>Gallus gallus</em>)
chromosome 33. Note that your VCF must be pre-filtered to only include
variant sites. More information on how to do this can be found in the
help page (<code><a href="../reference/read_vcf.html">?read_vcf</a></code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcf_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"galGal6-chr33.vcf.gz"</span>, package<span class="op">=</span><span class="st">"tfboot"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_vcf.html">read_vcf</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">vcf_file</span>, bsgenome<span class="op">=</span><span class="va">BSgenome.Ggallus.UCSC.galGal6</span><span class="op">)</span></span>
<span><span class="va">snps</span></span>
<span><span class="co">#&gt; GRanges object with 20000 ranges and 6 metadata columns:</span></span>
<span><span class="co">#&gt;                     seqnames    ranges strand | paramRangeID            REF</span></span>
<span><span class="co">#&gt;                        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;</span></span>
<span><span class="co">#&gt;     chr33:10054_G/T    chr33     10054      * |           NA              G</span></span>
<span><span class="co">#&gt;     chr33:11031_T/C    chr33     11031      * |           NA              T</span></span>
<span><span class="co">#&gt;     chr33:11148_T/G    chr33     11148      * |           NA              T</span></span>
<span><span class="co">#&gt;     chr33:11400_G/C    chr33     11400      * |           NA              G</span></span>
<span><span class="co">#&gt;     chr33:11496_G/A    chr33     11496      * |           NA              G</span></span>
<span><span class="co">#&gt;                 ...      ...       ...    ... .          ...            ...</span></span>
<span><span class="co">#&gt;   chr33:7810067_A/C    chr33   7810067      * |           NA              A</span></span>
<span><span class="co">#&gt;   chr33:7810166_G/A    chr33   7810166      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:7810888_T/G    chr33   7810888      * |           NA              T</span></span>
<span><span class="co">#&gt;   chr33:7810896_T/A    chr33   7810896      * |           NA              T</span></span>
<span><span class="co">#&gt;   chr33:7811431_G/A    chr33   7811431      * |           NA              G</span></span>
<span><span class="co">#&gt;                                ALT      QUAL      FILTER            SNP_id</span></span>
<span><span class="co">#&gt;                     &lt;DNAStringSet&gt; &lt;numeric&gt; &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt;     chr33:10054_G/T              T        NA           .   chr33:10054_G/T</span></span>
<span><span class="co">#&gt;     chr33:11031_T/C              C        NA           .   chr33:11031_T/C</span></span>
<span><span class="co">#&gt;     chr33:11148_T/G              G        NA           .   chr33:11148_T/G</span></span>
<span><span class="co">#&gt;     chr33:11400_G/C              C        NA           .   chr33:11400_G/C</span></span>
<span><span class="co">#&gt;     chr33:11496_G/A              A        NA           .   chr33:11496_G/A</span></span>
<span><span class="co">#&gt;                 ...            ...       ...         ...               ...</span></span>
<span><span class="co">#&gt;   chr33:7810067_A/C              C        NA           . chr33:7810067_A/C</span></span>
<span><span class="co">#&gt;   chr33:7810166_G/A              A        NA           . chr33:7810166_G/A</span></span>
<span><span class="co">#&gt;   chr33:7810888_T/G              G        NA           . chr33:7810888_T/G</span></span>
<span><span class="co">#&gt;   chr33:7810896_T/A              A        NA           . chr33:7810896_T/A</span></span>
<span><span class="co">#&gt;   chr33:7811431_G/A              A        NA           . chr33:7811431_G/A</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>The <code><a href="../reference/get_upstream_snps.html">get_upstream_snps()</a></code> function from tfboot takes in a
SNP list and a TxDb object, and turns intervals 5kb upstream of the
transcription start site for all genes in the list. See the help for
<code>?get_upstream_snps()</code> and the helper function it calls,
<code><a href="../reference/get_upstream.html">get_upstream()</a></code>. This function also respects strand
information embedded in the genome. For an example on building a TxDb
object from a GTF, see
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbPackage.html" class="external-link">?GenomicFeatures::makeTxDbPackage</a></code>.</p>
<p>Here we get the 5kb upstream promoter region of <em>all</em> genes
with <em>any</em> SNPs (limited to chromosome 33 in this example).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prosnps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_upstream_snps.html">get_upstream_snps</a></span><span class="op">(</span><span class="va">snps</span>, txdb<span class="op">=</span><span class="va">TxDb.Ggallus.UCSC.galGal6.refGene</span><span class="op">)</span></span>
<span></span>
<span><span class="va">prosnps</span></span>
<span><span class="co">#&gt; GRanges object with 685 ranges and 8 metadata columns:</span></span>
<span><span class="co">#&gt;                     seqnames    ranges strand | paramRangeID            REF</span></span>
<span><span class="co">#&gt;                        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;</span></span>
<span><span class="co">#&gt;   chr33:3630733_A/G    chr33   3630733      * |           NA              A</span></span>
<span><span class="co">#&gt;   chr33:3630900_A/T    chr33   3630900      * |           NA              A</span></span>
<span><span class="co">#&gt;   chr33:3631029_G/T    chr33   3631029      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:3631819_G/C    chr33   3631819      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:3631916_T/C    chr33   3631916      * |           NA              T</span></span>
<span><span class="co">#&gt;                 ...      ...       ...    ... .          ...            ...</span></span>
<span><span class="co">#&gt;   chr33:7721873_G/A    chr33   7721873      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:7721994_A/T    chr33   7721994      * |           NA              A</span></span>
<span><span class="co">#&gt;   chr33:7722028_G/T    chr33   7722028      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:7722600_C/A    chr33   7722600      * |           NA              C</span></span>
<span><span class="co">#&gt;   chr33:7723117_A/C    chr33   7723117      * |           NA              A</span></span>
<span><span class="co">#&gt;                                ALT      QUAL      FILTER            SNP_id</span></span>
<span><span class="co">#&gt;                     &lt;DNAStringSet&gt; &lt;numeric&gt; &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt;   chr33:3630733_A/G              G        NA           . chr33:3630733_A/G</span></span>
<span><span class="co">#&gt;   chr33:3630900_A/T              T        NA           . chr33:3630900_A/T</span></span>
<span><span class="co">#&gt;   chr33:3631029_G/T              T        NA           . chr33:3631029_G/T</span></span>
<span><span class="co">#&gt;   chr33:3631819_G/C              C        NA           . chr33:3631819_G/C</span></span>
<span><span class="co">#&gt;   chr33:3631916_T/C              C        NA           . chr33:3631916_T/C</span></span>
<span><span class="co">#&gt;                 ...            ...       ...         ...               ...</span></span>
<span><span class="co">#&gt;   chr33:7721873_G/A              A        NA           . chr33:7721873_G/A</span></span>
<span><span class="co">#&gt;   chr33:7721994_A/T              T        NA           . chr33:7721994_A/T</span></span>
<span><span class="co">#&gt;   chr33:7722028_G/T              T        NA           . chr33:7722028_G/T</span></span>
<span><span class="co">#&gt;   chr33:7722600_C/A              A        NA           . chr33:7722600_C/A</span></span>
<span><span class="co">#&gt;   chr33:7723117_A/C              C        NA           . chr33:7723117_A/C</span></span>
<span><span class="co">#&gt;                         gene_id ranges_original</span></span>
<span><span class="co">#&gt;                     &lt;character&gt;       &lt;IRanges&gt;</span></span>
<span><span class="co">#&gt;   chr33:3630733_A/G      776487 3635699-3637068</span></span>
<span><span class="co">#&gt;   chr33:3630900_A/T      776487 3635699-3637068</span></span>
<span><span class="co">#&gt;   chr33:3631029_G/T      776487 3635699-3637068</span></span>
<span><span class="co">#&gt;   chr33:3631819_G/C      776487 3635699-3637068</span></span>
<span><span class="co">#&gt;   chr33:3631916_T/C      776487 3635699-3637068</span></span>
<span><span class="co">#&gt;                 ...         ...             ...</span></span>
<span><span class="co">#&gt;   chr33:7721873_G/A      424136 7724062-7729597</span></span>
<span><span class="co">#&gt;   chr33:7721994_A/T      424136 7724062-7729597</span></span>
<span><span class="co">#&gt;   chr33:7722028_G/T      424136 7724062-7729597</span></span>
<span><span class="co">#&gt;   chr33:7722600_C/A      424136 7724062-7729597</span></span>
<span><span class="co">#&gt;   chr33:7723117_A/C      424136 7724062-7729597</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from galGal6 genome</span></span></code></pre></div>
<p>Now, let’s select a set of five random genes of interest.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">mygenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">prosnps</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">mygenes</span></span>
<span><span class="co">#&gt; [1] "693249"    "408041"    "100316005" "408042"    "100498692"</span></span></code></pre></div>
<p>What are those genes?</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">org.Gg.eg.db</span>, </span>
<span>                      key<span class="op">=</span><span class="va">mygenes</span>, </span>
<span>                      columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENTREZID"</span>, <span class="st">"SYMBOL"</span>, <span class="st">"GENENAME"</span><span class="op">)</span>, </span>
<span>                      keytype<span class="op">=</span><span class="st">"ENTREZID"</span><span class="op">)</span></span>
<span><span class="co">#&gt;    ENTREZID  SYMBOL      GENENAME</span></span>
<span><span class="co">#&gt; 1    693249   CAPN1     calpain 1</span></span>
<span><span class="co">#&gt; 2    408041   KRT6A    keratin 6A</span></span>
<span><span class="co">#&gt; 3 100316005 MIR2130 microRNA 2130</span></span>
<span><span class="co">#&gt; 4    408042   KRT75    keratin 75</span></span>
<span><span class="co">#&gt; 5 100498692 MIR3534 microRNA 3534</span></span></code></pre></div>
<p>We can then pull out SNPs in the 5kb promoter region of just those
SNPs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myprosnps</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">prosnps</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">mygenes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">myprosnps</span></span>
<span><span class="co">#&gt; GRanges object with 60 ranges and 8 metadata columns:</span></span>
<span><span class="co">#&gt;                     seqnames    ranges strand | paramRangeID            REF</span></span>
<span><span class="co">#&gt;                        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;</span></span>
<span><span class="co">#&gt;   chr33:3804056_G/C    chr33   3804056      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:3805103_A/C    chr33   3805103      * |           NA              A</span></span>
<span><span class="co">#&gt;   chr33:3805111_G/T    chr33   3805111      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:3805659_T/C    chr33   3805659      * |           NA              T</span></span>
<span><span class="co">#&gt;   chr33:3805709_C/G    chr33   3805709      * |           NA              C</span></span>
<span><span class="co">#&gt;                 ...      ...       ...    ... .          ...            ...</span></span>
<span><span class="co">#&gt;   chr33:7473031_T/C    chr33   7473031      * |           NA              T</span></span>
<span><span class="co">#&gt;   chr33:7473583_G/A    chr33   7473583      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:7473834_G/T    chr33   7473834      * |           NA              G</span></span>
<span><span class="co">#&gt;   chr33:7473857_T/C    chr33   7473857      * |           NA              T</span></span>
<span><span class="co">#&gt;   chr33:7475206_A/C    chr33   7475206      * |           NA              A</span></span>
<span><span class="co">#&gt;                                ALT      QUAL      FILTER            SNP_id</span></span>
<span><span class="co">#&gt;                     &lt;DNAStringSet&gt; &lt;numeric&gt; &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt;   chr33:3804056_G/C              C        NA           . chr33:3804056_G/C</span></span>
<span><span class="co">#&gt;   chr33:3805103_A/C              C        NA           . chr33:3805103_A/C</span></span>
<span><span class="co">#&gt;   chr33:3805111_G/T              T        NA           . chr33:3805111_G/T</span></span>
<span><span class="co">#&gt;   chr33:3805659_T/C              C        NA           . chr33:3805659_T/C</span></span>
<span><span class="co">#&gt;   chr33:3805709_C/G              G        NA           . chr33:3805709_C/G</span></span>
<span><span class="co">#&gt;                 ...            ...       ...         ...               ...</span></span>
<span><span class="co">#&gt;   chr33:7473031_T/C              C        NA           . chr33:7473031_T/C</span></span>
<span><span class="co">#&gt;   chr33:7473583_G/A              A        NA           . chr33:7473583_G/A</span></span>
<span><span class="co">#&gt;   chr33:7473834_G/T              T        NA           . chr33:7473834_G/T</span></span>
<span><span class="co">#&gt;   chr33:7473857_T/C              C        NA           . chr33:7473857_T/C</span></span>
<span><span class="co">#&gt;   chr33:7475206_A/C              C        NA           . chr33:7475206_A/C</span></span>
<span><span class="co">#&gt;                         gene_id ranges_original</span></span>
<span><span class="co">#&gt;                     &lt;character&gt;       &lt;IRanges&gt;</span></span>
<span><span class="co">#&gt;   chr33:3804056_G/C   100498692 3808915-3808996</span></span>
<span><span class="co">#&gt;   chr33:3805103_A/C   100498692 3808915-3808996</span></span>
<span><span class="co">#&gt;   chr33:3805111_G/T   100498692 3808915-3808996</span></span>
<span><span class="co">#&gt;   chr33:3805659_T/C   100498692 3808915-3808996</span></span>
<span><span class="co">#&gt;   chr33:3805709_C/G   100498692 3808915-3808996</span></span>
<span><span class="co">#&gt;                 ...         ...             ...</span></span>
<span><span class="co">#&gt;   chr33:7473031_T/C   100316005 7470778-7470842</span></span>
<span><span class="co">#&gt;   chr33:7473583_G/A   100316005 7470778-7470842</span></span>
<span><span class="co">#&gt;   chr33:7473834_G/T   100316005 7470778-7470842</span></span>
<span><span class="co">#&gt;   chr33:7473857_T/C   100316005 7470778-7470842</span></span>
<span><span class="co">#&gt;   chr33:7475206_A/C   100316005 7470778-7470842</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from galGal6 genome</span></span></code></pre></div>
<p>How many SNPs in the promoter region of each of my genes of
interest?</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">myprosnps</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 100316005 100498692    408041    408042    693249 </span></span>
<span><span class="co">#&gt;        11        15        12         7        15</span></span></code></pre></div>
<p>What’s the median number of SNPs across <em>all</em> genes?</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">prosnps</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 12.5</span></span></code></pre></div>
<p>Now, let’s run motifbreakR analysis with the 2022 JASPAR database on
those 5 genes. First let’s look at the TFBS motifs. See
<code><a href="https://rdrr.io/pkg/MotifDb/man/MotifDb.html" class="external-link">?MotifDb</a></code> for more info.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">MotifDb</span>, <span class="va">dataSource</span><span class="op">==</span><span class="st">"jaspar2022"</span><span class="op">)</span></span>
<span><span class="va">motifs</span></span>
<span><span class="co">#&gt; MotifDb object of length 1956</span></span>
<span><span class="co">#&gt; | Created from downloaded public sources, last update: 2022-Mar-04</span></span>
<span><span class="co">#&gt; | 1956 position frequency matrices from 1 source:</span></span>
<span><span class="co">#&gt; |         jaspar2022: 1956</span></span>
<span><span class="co">#&gt; | 19 organism/s</span></span>
<span><span class="co">#&gt; |           Hsapiens:  691</span></span>
<span><span class="co">#&gt; |          Athaliana:  568</span></span>
<span><span class="co">#&gt; |        Scerevisiae:  170</span></span>
<span><span class="co">#&gt; |      Dmelanogaster:  150</span></span>
<span><span class="co">#&gt; |                 NA:  144</span></span>
<span><span class="co">#&gt; |          Mmusculus:  143</span></span>
<span><span class="co">#&gt; |              other:   90</span></span>
<span><span class="co">#&gt; Mmusculus-jaspar2022-Arnt-MA0004.1 </span></span>
<span><span class="co">#&gt; Mmusculus-jaspar2022-Ahr::Arnt-MA0006.1 </span></span>
<span><span class="co">#&gt; Dmelanogaster-jaspar2022-br-MA0010.1 </span></span>
<span><span class="co">#&gt; Dmelanogaster-jaspar2022-br-MA0011.1 </span></span>
<span><span class="co">#&gt; Dmelanogaster-jaspar2022-br-MA0012.1 </span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; Athaliana-jaspar2022-HSFC1-MA1667.2 </span></span>
<span><span class="co">#&gt; Athaliana-jaspar2022-NAC017-MA1674.2 </span></span>
<span><span class="co">#&gt; Athaliana-jaspar2022-NAC062-MA1676.2 </span></span>
<span><span class="co">#&gt; Athaliana-jaspar2022-NTL8-MA1678.2 </span></span>
<span><span class="co">#&gt; Athaliana-jaspar2022-ERF11-MA1001.3</span></span></code></pre></div>
<p>Now, let’s run motifbreakR on these SNPs in the promoter regions of
these 5 genes. The result of the motifbreakR analysis is a GenomicRanges
object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbresGR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifbreakR/man/motifbreakR.html" class="external-link">motifbreakR</a></span><span class="op">(</span><span class="va">myprosnps</span>, pwmList<span class="op">=</span><span class="va">motifs</span><span class="op">)</span></span></code></pre></div>
<p>Additionally, let’s run motifbreakR on <em>all</em> SNPs for promoter
regions in <em>all</em> genes (remember, this entire analysis is
restricted to chromosome 33 only).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mballGR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifbreakR/man/motifbreakR.html" class="external-link">motifbreakR</a></span><span class="op">(</span><span class="va">prosnps</span>, pwmList<span class="op">=</span><span class="va">motifs</span><span class="op">)</span></span></code></pre></div>
<p>Let’s use tfboot’s <code><a href="../reference/mb_to_tibble.html">mb_to_tibble()</a></code> function to create a
compact tibble for each of these. We’ll use this data for our
statistical analysis coming up next.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbres</span> <span class="op">&lt;-</span> <span class="va">mbresGR</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/mb_to_tibble.html">mb_to_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">mball</span> <span class="op">&lt;-</span> <span class="va">mballGR</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/mb_to_tibble.html">mb_to_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbres</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,335 × 10</span></span></span>
<span><span class="co">#&gt;    gene_id   SNP_id      tf    pctRef pctAlt scoreRef scoreAlt effect alleleDiff</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 100316005 chr33:7472… ACE2   0.986  0.783     4.85     3.87 strong     -<span style="color: #BB0000;">0.980</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 100316005 chr33:7472… ADR1   0.759  0.917     3.34     4.01 weak        0.673</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 100316005 chr33:7471… AFT2   0.964  0.878     5.50     5.02 weak       -<span style="color: #BB0000;">0.480</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 100316005 chr33:7472… AFT2   0.902  0.740     5.15     4.24 strong     -<span style="color: #BB0000;">0.911</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 100316005 chr33:7473… AFT2   0.879  0.705     5.02     4.04 strong     -<span style="color: #BB0000;">0.980</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 100316005 chr33:7472… AGL42  0.859  0.693     5.14     4.15 strong     -<span style="color: #BB0000;">0.986</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 100316005 chr33:7471… AGL55  0.695  0.861     3.97     4.91 strong      0.939</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 100316005 chr33:7472… ALX3   0.972  0.776     4.23     3.43 strong     -<span style="color: #BB0000;">0.806</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 100316005 chr33:7473… ALX3   0.919  0.773     4.02     3.42 weak       -<span style="color: #BB0000;">0.600</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 100316005 chr33:7473… ARG80  0.731  0.961     3.16     4.15 strong      0.993</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,325 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: alleleEffectSize &lt;dbl&gt;</span></span></span>
<span><span class="va">mball</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12,689 × 10</span></span></span>
<span><span class="co">#&gt;    gene_id   SNP_id      tf    pctRef pctAlt scoreRef scoreAlt effect alleleDiff</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 100315917 chr33:5706… ARG80  0.979  0.749     4.23     3.24 strong     -<span style="color: #BB0000;">0.993</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 100315917 chr33:5706… ARG81  0.991  0.825     5.89     4.90 strong     -<span style="color: #BB0000;">0.982</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 100315917 chr33:5703… ARGFX  0.781  0.889     5.06     5.74 weak        0.677</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 100315917 chr33:5706… ARR1   0.634  0.880     2.53     3.42 strong      0.892</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 100315917 chr33:5707… AT1G…  0.882  0.779     7.60     6.73 strong     -<span style="color: #BB0000;">0.861</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 100315917 chr33:5704… AT3G…  0.784  0.892     5.13     5.79 weak        0.666</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 100315917 chr33:5706… ATHB…  0.781  0.890     4.12     4.67 weak        0.552</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 100315917 chr33:5706… ATHB…  0.721  0.868     4.98     5.98 strong      0.993</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 100315917 chr33:5706… ATHB…  0.813  0.966     5.32     6.32 strong      0.998</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 100315917 chr33:5706… ATHB…  0.778  0.931     4.98     5.94 strong      0.963</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 12,679 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: alleleEffectSize &lt;dbl&gt;</span></span></span></code></pre></div>
<p>In reality, when you perform the motifbreakR results across the whole
genome, the analysis will take some time (perhaps hours), depending on
the number of cores allocated to the job. I recommend that once you
perform this once for every gene in the genome, you save the compact
tibble (<code>mball</code> in this example) using <code><a href="https://rdrr.io/r/base/save.html" class="external-link">save()</a></code> or
<code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS()</a></code> to read in later, avoiding the need to re-run the
genome-wide motifbreakR analysis each time. You can do this with
something like this:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save the data so you don't have to run this long-running analysis again</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">mball</span>, file<span class="op">=</span><span class="st">"genome-wide-motifbreakr-results.rds"</span><span class="op">)</span></span>
<span><span class="co"># Read them back in from a file</span></span>
<span><span class="va">mball</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"genome-wide-motifbreakr-results.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="assessing-statistical-signficance">Assessing statistical signficance<a class="anchor" aria-label="anchor" href="#assessing-statistical-signficance"></a>
</h2>
<div class="section level3">
<h3 id="motivation-1">Motivation<a class="anchor" aria-label="anchor" href="#motivation-1"></a>
</h3>
<p>In our example above we have a set of five genes. We’ve run
motifbreakR on SNPs in the 5kb upstream region of those genes, and we
want to ask questions such as “are there statistically more SNPs in the
promoters of <em>these</em> five genes than a random selection of five
genes?” or “do SNPs in the promoters of <em>these</em> five genes
statistically enriched for those that disrupt TFBS, more so than
randomly selected genes?” The tfboot package can help answer these
questions.</p>
<p>We initially ran motifbreakR on our set of five genes of interest. To
address the kinds of questions above, we could go back to the set of all
genes (again, here just those limited to chromosome 33 for demonstration
purposes), and randomly draw 1000 sets of 5 genes, run motifbreakR on
SNPs in those promoters, then compare our stats to the empirical null
distribution to obtain a p-value. However, running motifbreakR on five
genes takes time. Imagine our gene set was 300 genes. Then we would need
to run motifbreakR 300,000 times to generate our null distribution.
Further, if we had <em>multiple</em> gene sets, e.g., one with 100
genes, another with 500 genes, another with 400 genes, we would need to
run motifbreakR 1,000,000 times to generate null distrubitions of 1,000
bootstraps for each of these intervals.</p>
<p>Instead, we can generate the motifbreakR results for <em>all</em>
genes, do this only once, then save those results to file, as we’ve done
above. Then, when we need to compute our bootstrap sampling to create
the empirical null distribution, we only have to bootstrap precomputed
results, rather than bootstrapping the motifbreakR procedure itself.</p>
</div>
<div class="section level3">
<h3 id="demonstration">Demonstration<a class="anchor" aria-label="anchor" href="#demonstration"></a>
</h3>
<p>Let’s look again at our motifbreakR results on our genes of
interest:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Look at the data</span></span>
<span><span class="va">mbres</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,335 × 10</span></span></span>
<span><span class="co">#&gt;    gene_id   SNP_id      tf    pctRef pctAlt scoreRef scoreAlt effect alleleDiff</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 100316005 chr33:7472… ACE2   0.986  0.783     4.85     3.87 strong     -<span style="color: #BB0000;">0.980</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 100316005 chr33:7472… ADR1   0.759  0.917     3.34     4.01 weak        0.673</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 100316005 chr33:7471… AFT2   0.964  0.878     5.50     5.02 weak       -<span style="color: #BB0000;">0.480</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 100316005 chr33:7472… AFT2   0.902  0.740     5.15     4.24 strong     -<span style="color: #BB0000;">0.911</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 100316005 chr33:7473… AFT2   0.879  0.705     5.02     4.04 strong     -<span style="color: #BB0000;">0.980</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 100316005 chr33:7472… AGL42  0.859  0.693     5.14     4.15 strong     -<span style="color: #BB0000;">0.986</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 100316005 chr33:7471… AGL55  0.695  0.861     3.97     4.91 strong      0.939</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 100316005 chr33:7472… ALX3   0.972  0.776     4.23     3.43 strong     -<span style="color: #BB0000;">0.806</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 100316005 chr33:7473… ALX3   0.919  0.773     4.02     3.42 weak       -<span style="color: #BB0000;">0.600</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 100316005 chr33:7473… ARG80  0.731  0.961     3.16     4.15 strong      0.993</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,325 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: alleleEffectSize &lt;dbl&gt;</span></span></span>
<span><span class="co"># How many genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mbres</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 5</span></span>
<span><span class="co"># How many TFBS are disrupted for each gene?</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">mbres</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 100316005 100498692    408041    408042    693249 </span></span>
<span><span class="co">#&gt;       436       206       361       131       201</span></span>
<span><span class="co"># How many "strong" in each gene?</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">mbres</span><span class="op">$</span><span class="va">gene_id</span>, <span class="va">mbres</span><span class="op">$</span><span class="va">effect</span><span class="op">)</span></span>
<span><span class="co">#&gt;            </span></span>
<span><span class="co">#&gt;             strong weak</span></span>
<span><span class="co">#&gt;   100316005    326  110</span></span>
<span><span class="co">#&gt;   100498692    152   54</span></span>
<span><span class="co">#&gt;   408041       225  136</span></span>
<span><span class="co">#&gt;   408042       101   30</span></span>
<span><span class="co">#&gt;   693249       146   55</span></span></code></pre></div>
<p>Now, let’s review the motifbreakR on <em>all</em> genes of
interest</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Look at the data (note the number of rows)</span></span>
<span><span class="va">mball</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12,689 × 10</span></span></span>
<span><span class="co">#&gt;    gene_id   SNP_id      tf    pctRef pctAlt scoreRef scoreAlt effect alleleDiff</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 100315917 chr33:5706… ARG80  0.979  0.749     4.23     3.24 strong     -<span style="color: #BB0000;">0.993</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 100315917 chr33:5706… ARG81  0.991  0.825     5.89     4.90 strong     -<span style="color: #BB0000;">0.982</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 100315917 chr33:5703… ARGFX  0.781  0.889     5.06     5.74 weak        0.677</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 100315917 chr33:5706… ARR1   0.634  0.880     2.53     3.42 strong      0.892</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 100315917 chr33:5707… AT1G…  0.882  0.779     7.60     6.73 strong     -<span style="color: #BB0000;">0.861</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 100315917 chr33:5704… AT3G…  0.784  0.892     5.13     5.79 weak        0.666</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 100315917 chr33:5706… ATHB…  0.781  0.890     4.12     4.67 weak        0.552</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 100315917 chr33:5706… ATHB…  0.721  0.868     4.98     5.98 strong      0.993</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 100315917 chr33:5706… ATHB…  0.813  0.966     5.32     6.32 strong      0.998</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 100315917 chr33:5706… ATHB…  0.778  0.931     4.98     5.94 strong      0.963</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 12,679 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: alleleEffectSize &lt;dbl&gt;</span></span></span>
<span><span class="co"># How many genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mball</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 56</span></span>
<span><span class="co"># How many TFBS are disrupted for each gene? Just show the top 5</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">mball</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span>decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 100316005 102466833    408041    429501 100859133 </span></span>
<span><span class="co">#&gt;       436       427       361       360       343</span></span></code></pre></div>
<p>The tfboot function <code>mb_summarize</code> will summarize the
results from a motifbreakR analysis into a single-row table. See help
for <code><a href="../reference/mb_summarize.html">?mb_summarize</a></code> for details. This returns a table with
the following columns:</p>
<ol style="list-style-type: decimal">
<li>
<code>ngenes</code>: The number of genes in the SNP set.</li>
<li>
<code>nsnps</code>: The number of SNPs total.</li>
<li>
<code>nstrong</code>: The number of SNPs with a “strong”
effect.</li>
<li>
<code>alleleDiffAbsMean</code> The mean of the absolute values of
the <code>alleleDiff</code> scores.</li>
<li>
<code>alleleDiffAbsSum</code> The sum of the absolute values of the
<code>alleleDiff</code> scores.</li>
<li>
<code>alleleEffectSizeAbsMean</code> The mean of the absolute values
of the <code>alleleEffectSize</code> scores.</li>
<li>
<code>alleleEffectSizeAbsSum</code> The sum of the absolute values
of the <code>alleleEffectSize</code> scores.</li>
</ol>
<p>Let’s run it on our set of genes of interest:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbsmry</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mb_summarize.html">mb_summarize</a></span><span class="op">(</span><span class="va">mbres</span><span class="op">)</span></span>
<span><span class="va">mbsmry</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 7</span></span></span>
<span><span class="co">#&gt;   ngenes nsnps nstrong alleleDiffAbsMean alleleDiffAbsSum alleleEffectSizeAbsM…¹</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      5  <span style="text-decoration: underline;">1</span>335     950             0.798            <span style="text-decoration: underline;">1</span>066.                  0.155</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​alleleEffectSizeAbsMean</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: alleleEffectSizeAbsSum &lt;dbl&gt;</span></span></span></code></pre></div>
<p>So, how do these numbers compare to, for example, the number of SNPs,
number of strong effects, the average absolute allele difference, or
average absolute effect size, from a set of SNPs in promoters in a
randomly chosen sample of the same number of genes?</p>
<p>We can use the <code>mb_bootstrap</code> function to generate an
empirical null distribution given the precomputed set of motifbreakR for
the universe of all genes. This function takes motifbreakR results as a
tibble (from <code>mb_to_tibble</code>), draws <code>boots</code> random
samples of <code>ngenes</code> genes, and returns as a list (1) a wide
tibble with results for each bootstrap, and (2) another tibble with the
distribution of each metric as a column. We set the random number
generator for reproducibility, and we limit the number of bootstrap
resamples here for speed.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">mbboot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mb_bootstrap.html">mb_bootstrap</a></span><span class="op">(</span><span class="va">mball</span>, ngenes<span class="op">=</span><span class="fl">5</span>, boots<span class="op">=</span><span class="fl">250</span><span class="op">)</span></span>
<span><span class="va">mbboot</span><span class="op">$</span><span class="va">bootwide</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 250 × 9</span></span></span>
<span><span class="co">#&gt;     boot genes           ngenes nsnps nstrong alleleDiffAbsMean alleleDiffAbsSum</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 431304;426469;…      5  <span style="text-decoration: underline;">1</span>062     750             0.797             847.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2 426183;395959;…      5  <span style="text-decoration: underline;">1</span>244     914             0.808            <span style="text-decoration: underline;">1</span>005.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3 102465361;4261…      5   886     655             0.812             720.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4 396007;407779;…      5  <span style="text-decoration: underline;">1</span>514    <span style="text-decoration: underline;">1</span>139             0.815            <span style="text-decoration: underline;">1</span>234.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5 426883;396544;…      5   925     623             0.787             728.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6 425059;429035;…      5  <span style="text-decoration: underline;">1</span>350    <span style="text-decoration: underline;">1</span>026             0.821            <span style="text-decoration: underline;">1</span>109.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7 408042;426880;…      5   823     627             0.823             677.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8 396170;425058;…      5  <span style="text-decoration: underline;">1</span>262     910             0.805            <span style="text-decoration: underline;">1</span>016.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9 102466833;4261…      5  <span style="text-decoration: underline;">1</span>263     963             0.822            <span style="text-decoration: underline;">1</span>038.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10 429035;408042;…      5  <span style="text-decoration: underline;">1</span>289     987             0.823            <span style="text-decoration: underline;">1</span>061.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 240 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: alleleEffectSizeAbsMean &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   alleleEffectSizeAbsSum &lt;dbl&gt;</span></span></span>
<span><span class="va">mbboot</span><span class="op">$</span><span class="va">bootdist</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   metric                  bootdist   </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> alleleDiffAbsMean       <span style="color: #949494;">&lt;dbl [250]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> alleleDiffAbsSum        <span style="color: #949494;">&lt;dbl [250]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> alleleEffectSizeAbsMean <span style="color: #949494;">&lt;dbl [250]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> alleleEffectSizeAbsSum  <span style="color: #949494;">&lt;dbl [250]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> nsnps                   <span style="color: #949494;">&lt;dbl [250]&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> nstrong                 <span style="color: #949494;">&lt;dbl [250]&gt;</span></span></span></code></pre></div>
<p>Finally, we can use the <code><a href="../reference/mb_bootstats.html">mb_bootstats()</a></code> function to take
in the summary on <em>our</em> genes of interest (output from
<code>mb_summary()</code>) and the bootstrapped empirical null
distribution (from <code><a href="../reference/mb_bootstrap.html">mb_bootstrap()</a></code> run on motifbreakR
results from <em>all</em> genes), to compare our critical values (in
<code>mbsmry</code> here) to our empirical null distribution (called
<code>mbboot</code> here).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bootstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mb_bootstats.html">mb_bootstats</a></span><span class="op">(</span>mbsmry <span class="op">=</span> <span class="va">mbsmry</span>, mbboot <span class="op">=</span> <span class="va">mbboot</span><span class="op">)</span></span>
<span><span class="va">bootstats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   metric                      stat bootdist     bootmax     p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> nsnps                   <span style="text-decoration: underline;">1</span>335     <span style="color: #949494;">&lt;dbl [250]&gt;</span> <span style="text-decoration: underline;">1</span>597     0.136</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> nstrong                  950     <span style="color: #949494;">&lt;dbl [250]&gt;</span> <span style="text-decoration: underline;">1</span>210     0.188</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> alleleDiffAbsMean          0.798 <span style="color: #949494;">&lt;dbl [250]&gt;</span>    0.831 0.836</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> alleleDiffAbsSum        <span style="text-decoration: underline;">1</span>066.    <span style="color: #949494;">&lt;dbl [250]&gt;</span> <span style="text-decoration: underline;">1</span>304.    0.148</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> alleleEffectSizeAbsMean    0.155 <span style="color: #949494;">&lt;dbl [250]&gt;</span>    0.163 0.456</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> alleleEffectSizeAbsSum   206.    <span style="color: #949494;">&lt;dbl [250]&gt;</span>  251.    0.128</span></span></code></pre></div>
<p>We can also visualize this graphically with the
<code><a href="../reference/plot_bootstats.html">plot_bootstats()</a></code> function, which takes the results from
<code><a href="../reference/mb_bootstats.html">mb_bootstats()</a></code> as input.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bootstats.html">plot_bootstats</a></span><span class="op">(</span><span class="va">bootstats</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<p>From these results we see that there is no meaningful statistical
enrichment of TFBS-disrupting SNPs in this set of genes compared to
randomly selected genes (this is expected – our five genes <em>were</em>
randomly chosen).</p>
</div>
<div class="section level3">
<h3 id="caveats">Caveats<a class="anchor" aria-label="anchor" href="#caveats"></a>
</h3>
<p><strong>The set of motifbreakR results for all genes must be
recomputed on each new sample.</strong> This vignette demonstrates how
to precompute the motifbreakR results <em>once</em> for the universe of
all genes. You can use this precomputed set to very quickly bootstrap an
empirical null distribution of a set of <em>k</em> randomly selected
genes. However, this background set is unique to each sample. Each
sample will have a different set of SNPs, and this precomputing
procedure computes motifbreakR results for SNPs in upstream regions of
all genes. Similarly, if the upstream region changes from the default
5,000 bp, results for all genes will need to be recomputed again.</p>
<p><strong>SNP density matters.</strong> Promoter regions with a larger
number of SNPs are going to have larger values for the sum of the
absolute values of the allele difference and effect sizes. This should
not affect the mean of these metrics.</p>
<p><strong>Genes, not transcripts.</strong> This vignette demonstrates
running motifbreakR on SNPs in 5kb upstream regions defined by
<em>genes</em> (see the help for <code><a href="../reference/get_upstream.html">?get_upstream</a></code> and
<code><a href="../reference/get_upstream_snps.html">?get_upstream_snps</a></code> and
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">?GenomicFeatures::genes</a></code>). This is using the most 5’
transcript as <em>the</em> transcription start site, and looking 5kb
upstream of that position. This does not consider alternative
transcription start sites. Further development is required to contend
with multiple sets of TFBS motif disruption scores for the same gene if
transcripts are to be used.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] tfboot_1.0.1                            </span></span>
<span><span class="co">#&gt;  [2] motifbreakR_2.16.0                      </span></span>
<span><span class="co">#&gt;  [3] MotifDb_1.44.0                          </span></span>
<span><span class="co">#&gt;  [4] plyranges_1.22.0                        </span></span>
<span><span class="co">#&gt;  [5] org.Gg.eg.db_3.18.0                     </span></span>
<span><span class="co">#&gt;  [6] TxDb.Ggallus.UCSC.galGal6.refGene_3.10.0</span></span>
<span><span class="co">#&gt;  [7] GenomicFeatures_1.54.3                  </span></span>
<span><span class="co">#&gt;  [8] AnnotationDbi_1.64.1                    </span></span>
<span><span class="co">#&gt;  [9] Biobase_2.62.0                          </span></span>
<span><span class="co">#&gt; [10] BSgenome.Ggallus.UCSC.galGal6_1.4.2     </span></span>
<span><span class="co">#&gt; [11] BSgenome_1.70.1                         </span></span>
<span><span class="co">#&gt; [12] rtracklayer_1.62.0                      </span></span>
<span><span class="co">#&gt; [13] BiocIO_1.12.0                           </span></span>
<span><span class="co">#&gt; [14] Biostrings_2.70.2                       </span></span>
<span><span class="co">#&gt; [15] XVector_0.42.0                          </span></span>
<span><span class="co">#&gt; [16] GenomicRanges_1.54.1                    </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.38.5                     </span></span>
<span><span class="co">#&gt; [18] IRanges_2.36.0                          </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.40.2                        </span></span>
<span><span class="co">#&gt; [20] BiocGenerics_0.48.1                     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7                filelock_1.0.3             </span></span>
<span><span class="co">#&gt;   [3] R.oo_1.26.0                 tibble_3.2.1               </span></span>
<span><span class="co">#&gt;   [5] XML_3.99-0.16.1             rpart_4.1.21               </span></span>
<span><span class="co">#&gt;   [7] DirichletMultinomial_1.44.0 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;   [9] lattice_0.21-9              ensembldb_2.26.0           </span></span>
<span><span class="co">#&gt;  [11] MASS_7.3-60                 backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3              Hmisc_5.1-1                </span></span>
<span><span class="co">#&gt;  [15] sass_0.4.8                  rmarkdown_2.25             </span></span>
<span><span class="co">#&gt;  [17] jquerylib_0.1.4             yaml_2.3.8                 </span></span>
<span><span class="co">#&gt;  [19] Gviz_1.46.1                 DBI_1.2.1                  </span></span>
<span><span class="co">#&gt;  [21] CNEr_1.38.0                 RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [23] ade4_1.7-22                 abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [25] zlibbioc_1.48.0             purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [27] R.utils_2.12.3              AnnotationFilter_1.26.0    </span></span>
<span><span class="co">#&gt;  [29] biovizBase_1.50.0           RCurl_1.98-1.14            </span></span>
<span><span class="co">#&gt;  [31] nnet_7.3-19                 pracma_2.4.4               </span></span>
<span><span class="co">#&gt;  [33] VariantAnnotation_1.48.1    rappdirs_0.3.3             </span></span>
<span><span class="co">#&gt;  [35] GenomeInfoDbData_1.2.11     seqLogo_1.68.0             </span></span>
<span><span class="co">#&gt;  [37] annotate_1.80.0             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [39] codetools_0.2-19            DelayedArray_0.28.0        </span></span>
<span><span class="co">#&gt;  [41] xml2_1.3.6                  tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [43] farver_2.1.1                matrixStats_1.2.0          </span></span>
<span><span class="co">#&gt;  [45] BiocFileCache_2.10.1        base64enc_0.1-3            </span></span>
<span><span class="co">#&gt;  [47] GenomicAlignments_1.38.2    jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;  [49] motifStack_1.46.0           Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [51] systemfonts_1.0.5           tools_4.3.2                </span></span>
<span><span class="co">#&gt;  [53] progress_1.2.3              ragg_1.2.7                 </span></span>
<span><span class="co">#&gt;  [55] TFMPvalue_0.0.9             Rcpp_1.0.12                </span></span>
<span><span class="co">#&gt;  [57] glue_1.7.0                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [59] SparseArray_1.2.3           xfun_0.41                  </span></span>
<span><span class="co">#&gt;  [61] MatrixGenerics_1.14.0       dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [63] withr_3.0.0                 fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [65] latticeExtra_0.6-30         fansi_1.0.6                </span></span>
<span><span class="co">#&gt;  [67] caTools_1.18.2              digest_0.6.34              </span></span>
<span><span class="co">#&gt;  [69] R6_2.5.1                    textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [71] colorspace_2.1-0            GO.db_3.18.0               </span></span>
<span><span class="co">#&gt;  [73] gtools_3.9.5                poweRlaw_0.80.0            </span></span>
<span><span class="co">#&gt;  [75] jpeg_0.1-10                 dichromat_2.0-0.1          </span></span>
<span><span class="co">#&gt;  [77] biomaRt_2.58.2              RSQLite_2.3.5              </span></span>
<span><span class="co">#&gt;  [79] R.methodsS3_1.8.2           tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [81] utf8_1.2.4                  generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [83] data.table_1.15.0           prettyunits_1.2.0          </span></span>
<span><span class="co">#&gt;  [85] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt;  [87] S4Arrays_1.2.0              TFBSTools_1.40.0           </span></span>
<span><span class="co">#&gt;  [89] pkgconfig_2.0.3             gtable_0.3.4               </span></span>
<span><span class="co">#&gt;  [91] blob_1.2.4                  htmltools_0.5.7            </span></span>
<span><span class="co">#&gt;  [93] ProtGenerics_1.34.0         scales_1.3.0               </span></span>
<span><span class="co">#&gt;  [95] png_0.1-8                   knitr_1.45                 </span></span>
<span><span class="co">#&gt;  [97] rstudioapi_0.15.0           tzdb_0.4.0                 </span></span>
<span><span class="co">#&gt;  [99] reshape2_1.4.4              rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [101] checkmate_2.3.1             curl_5.2.0                 </span></span>
<span><span class="co">#&gt; [103] cachem_1.0.8                stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [105] parallel_4.3.2              foreign_0.8-85             </span></span>
<span><span class="co">#&gt; [107] restfulr_0.0.15             desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [109] pillar_1.9.0                vctrs_0.6.5                </span></span>
<span><span class="co">#&gt; [111] dbplyr_2.4.0                xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [113] cluster_2.1.4               htmlTable_2.4.2            </span></span>
<span><span class="co">#&gt; [115] evaluate_0.23               readr_2.1.5                </span></span>
<span><span class="co">#&gt; [117] cli_3.6.2                   compiler_4.3.2             </span></span>
<span><span class="co">#&gt; [119] Rsamtools_2.18.0            rlang_1.1.3                </span></span>
<span><span class="co">#&gt; [121] crayon_1.5.2                labeling_0.4.3             </span></span>
<span><span class="co">#&gt; [123] interp_1.1-6                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt; [125] fs_1.6.3                    stringi_1.8.3              </span></span>
<span><span class="co">#&gt; [127] deldir_2.0-2                BiocParallel_1.36.0        </span></span>
<span><span class="co">#&gt; [129] munsell_0.5.0               lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [131] Matrix_1.6-1.1              hms_1.1.3                  </span></span>
<span><span class="co">#&gt; [133] bit64_4.0.5                 ggplot2_3.4.4              </span></span>
<span><span class="co">#&gt; [135] KEGGREST_1.42.0             highr_0.10                 </span></span>
<span><span class="co">#&gt; [137] SummarizedExperiment_1.32.0 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [139] bslib_0.6.1                 bit_4.0.5                  </span></span>
<span><span class="co">#&gt; [141] splitstackshape_1.4.8</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Stephen Turner.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
